### PR TITLE
Talk directly to systemd

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
Ouch, there was some mistake with the "Closes" lines.

File: pkg/base1/cockpit.js [hunk 1]
```diff
@@ -1757,6 +1757,11 @@ function full_scope(cockpit, $, po) {
             ensure_cache();
             return new DBusProxies(self, cache, String(iface), String(path_namespace));
         };
+
+        self.update_props = function update_props(path, iface, props) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: I would suggest:
 * Having this trigger the notify event correctly.
 * Using the same input data structure here as with msg.notify and the notify event.
 * Calling this method ```.notify()```

File: pkg/systemd/Makefile.am [hunk 2]
```diff
@@ -2,8 +2,9 @@ systemddir = $(pkgdatadir)/system
 systemd_DATA = \
 	pkg/systemd/bundle.min.js.gz \
 	pkg/systemd/log.min.html.gz \
-	pkg/systemd/manifest.json \
 	pkg/systemd/terminal.min.html.gz \
+	pkg/systemd/init.min.html.gz \
+	pkg/systemd/manifest.json \
```
Comment (stefwalter) [inline pkg/systemd/Makefile.am]: Alphabetical?

File: pkg/systemd/Makefile.am [hunk 3]
```diff
@@ -14,12 +15,15 @@ systemddebug_DATA = \
 	pkg/systemd/server.js \
 	pkg/systemd/terminal.js \
 	pkg/systemd/terminal.html \
+	pkg/systemd/init.js \
+	pkg/systemd/init.html \
```
Comment (stefwalter) [inline pkg/systemd/Makefile.am]: Alphabetical?

File: pkg/systemd/Makefile.am [hunk 4]
```diff
@@ -14,12 +15,15 @@ systemddebug_DATA = \
...
 	$(NULL)
 
 systemd_BUNDLE = \
 	pkg/systemd/log.min.js \
 	pkg/systemd/server.min.js \
 	pkg/systemd/terminal.min.js \
+	pkg/systemd/init.min.js \
```
Comment (stefwalter) [inline pkg/systemd/Makefile.am]: Alphabetical?

File: pkg/systemd/Makefile.am [hunk 5]
```diff
@@ -35,6 +39,7 @@ CLEANFILES += \
 	pkg/systemd/bundle.min.js \
 	pkg/systemd/log.min.html \
 	pkg/systemd/terminal.min.html \
+	pkg/systemd/init.min.html \
```
Comment (stefwalter) [inline pkg/systemd/Makefile.am]: Alphabetical?

File: pkg/systemd/Makefile.am [hunk 6]
```diff
@@ -42,6 +47,7 @@ EXTRA_DIST += \
...
```
Comment (stefwalter) [inline pkg/systemd/Makefile.am]: Alphabetical?

File: pkg/systemd/init.html [hunk 7]
```diff
@@ -0,0 +1,195 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <title>Services</title>
```
Comment (stefwalter) [inline pkg/systemd/init.html]: translatable?

File: pkg/systemd/init.html [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                {{Unit.LoadState}} ({{Unit.FragmentPath}}; {{Unit.UnitFileState}})
+                {{/HasLoadError}}
+              </td>
+              <td style="text-align:right">
+                {{{FileButton}}}
+              </td>
+            </tr>
+            {{#Template}}
+            <tr>
+              <td>This unit is an instance of the <a data-goto-unit="{{Template}}">{{.}}</a> template.</td>
```
Comment (stefwalter) [inline pkg/systemd/init.html]: Translatable?

File: pkg/systemd/init.html [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    </div>
+  </script>
+
+  <script id="service-template-tmpl" type="x-template/mustache">
+    <div class="panel panel-default">
+      <div class="panel-heading">{{Id}} Template</div>
+      <div class="list-group">
+        <div class="list-group-item">
+          <input style="width:50em"></input>
+          <button>Instantiate</button>
```
Comment (stefwalter) [inline pkg/systemd/init.html]: Translatable?

File: pkg/systemd/init.html [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+            {{/Template}}
+          </table>
+        </div>
+      </div>
+    </div>
+  </script>
+
+  <script id="service-template-tmpl" type="x-template/mustache">
+    <div class="panel panel-default">
+      <div class="panel-heading">{{Id}} Template</div>
```
Comment (stefwalter) [inline pkg/systemd/init.html]: Translatable?

File: pkg/systemd/init.html [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  <div class="modal" id="service-no-install-info-dialog"
+       tabindex="-1" role="dialog"
+       data-backdrop="static">
+    <div class="modal-dialog">
+      <div class="modal-content">
+        <div class="modal-header">
+          <h4 class="modal-title">Note</h4>
+        </div>
+        <div class="modal-body">
+          <p>This unit is not designed to be enabled explicitly.
```
Comment (stefwalter) [inline pkg/systemd/init.html]: Lots more untranslatable strings that we should probably mark? I won't call them all out individually.

File: pkg/systemd/init.js [hunk 12]
```diff
@@ -0,0 +1,745 @@
...
+     */
+
+    function systemd_escape(str) {
+
+        function name_esc(str) {
+            var validchars = /[0-9a-zA-Z:-_.\\]/;
+            var res = "";
+            var i;
+
+            for (i = 0; i < str.length; i++) {
```
Comment (stefwalter) [inline pkg/systemd/init.js]: If this code is used a *lot* we may want to assign str.length to a variable. If this code is used rarely, ignore my silliness.

File: pkg/systemd/init.js [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+                return name_esc(str);
+        }
+
+        if (str.length > 0 && str[0] == "/")
+            return path_esc(str);
+        else
+            return name_esc(str);
+    }
+
+    var systemd_client = cockpit.dbus("org.freedesktop.systemd1", { superuser: true });
```
Comment (stefwalter) [inline pkg/systemd/init.js]: The 'superuser: true' is only necessary on older systemd. Perhaps have a TODO here linking to our superuser optional bug?

File: pkg/systemd/init.js [hunk 14]
```diff
@@ -0,0 +1,745 @@
...
+            cur_journal_watcher.stop();
+            cur_journal_watcher = null;
+        }
+
+        function render() {
+            var unit_def;
+            var active_state = cur_unit.ActiveState;
+            if (active_state == 'active' || active_state == 'reloading' ||
+                active_state == 'activating')
+                unit_def = 1;
```
Comment (stefwalter) [inline pkg/systemd/init.js]: Comments after the number indicating what it represents?

File: pkg/systemd/init.js [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+            var load_state = cur_unit.LoadState;
+            var file_state = cur_unit.UnitFileState;
+            if (load_state == 'masked')
+                file_def = 7;
+            else if (file_state == 'static')
+                file_def = 5;
+            else if (file_state == 'enabled')
+                file_def = 2;
+            else
+                file_def = 0;
```
Comment (stefwalter) [inline pkg/systemd/init.js]: Comments after the number indicating what it represents ... and above?

File: pkg/systemd/init.js [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+            systemd_manager.Reload();
+        }
+
+        var parsed_action = $(this).attr("data-action").split(":");
+
+        var method = parsed_action[1];
+        var args;
+        var done;
+
+        if (parsed_action[0] == 'u') {
```
Comment (stefwalter) [inline pkg/systemd/init.js]: Comments here, seems a bit magical at first glance.

**Comment by stefwalter** [comment]
I get this error when navigating to the 'Services' page:

```
TypeError: systemd_manager.Subscribe is not a function
```

File: pkg/base1/cockpit.js (hunk 1 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Yes, done.  Docs are missing still.

File: pkg/systemd/init.html (hunk 11 above)
Comment (mvollmer) [inline pkg/systemd/init.html]: Yes, I was very sloppy there.

File: pkg/systemd/init.js (hunk 12 above)
Comment (mvollmer) [inline pkg/systemd/init.js]: It is used when instantiating a template, so less than once a month. :-)

File: pkg/base1/cockpit.js [hunk 17]
```diff
@@ -1611,6 +1602,21 @@ function full_scope(cockpit, $, po) {
...
+                        cache.remove(path, iface);
+                    else
+                        cache.update(path, iface, props);
+                });
+            });
+            $(self).triggerHandler("notify", [ data ]);
+        }
+
+        this.notify = notify;
+
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Ah yes, docs and tests.

File: pkg/systemd/init.js (hunk 13 above)
Comment (mvollmer) [inline pkg/systemd/init.js]: Yes, good idea: #2151.

**Comment by stefwalter** [comment]
Fixes #585

**Comment by stefwalter** [comment]
This pull request should remove the cgroup-show.h and cgroup-show.c files as well. They're no longer used.

File: pkg/systemd/init.js (hunk 14 above)
Comment (mvollmer) [inline pkg/systemd/init.js]: Done

File: pkg/systemd/init.js (hunk 15 above)
Comment (mvollmer) [inline pkg/systemd/init.js]: Done

File: pkg/systemd/init.js (hunk 16 above)
Comment (mvollmer) [inline pkg/systemd/init.js]: Rewritten.

File: pkg/base1/cockpit.js (hunk 17 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Done

**Comment by mvollmer** [comment]
> This pull request should remove the cgroup-show.h and cgroup-show.c files as well. They're no longer used.

Good point, done.

File: pkg/systemd/Makefile.am (hunk 2 above)
Comment (mvollmer) [inline pkg/systemd/Makefile.am]: All done.

**Comment by mvollmer** [comment]
> I get this error when navigating to the 'Services' page:

Can you reproduce this reliably?  This shouldn't happen(tm) unless your systemd does not have that method.

**Comment by stefwalter** [comment]
> Can you reproduce this reliably? This shouldn't happen(tm) unless your systemd does not have that method.

Do you wait for the proxy to become valid before calling Subscribe? If you don't want to do that, you should use ```proxy.call("Subscribe" ...)```

**Comment by mvollmer** [comment]
> Do you wait for the proxy to become valid before calling Subscribe? If you don't want to do that, you should use proxy.call("Subscribe" ...)

Yes:

```
     function init() {
        systemd_manager.wait(function () {
            systemd_manager.Subscribe().
                fail(function (error) {
                    console.log(error);
                });
            update();
        });
    }
```

**Comment by mvollmer** [comment]
check-servies on rhel-7 needs attention.

**Comment by mvollmer** [comment]
The "Since" timestamp is the epoch often.

**Comment by mvollmer** [comment]
The whole overview page is reconstructed incrementally when a service is enabled or disabled.  That's a bit distracting.

**Comment by stefwalter** [comment]
> The whole overview page is reconstructed incrementally when a service is enabled or disabled. That's a bit distracting.

In general, we need to get better at this throughout cockpit. It's worth investing in figuring out a way to do this "right"

@petervo might also have ideas ^^

**Comment by mvollmer** [comment]
> In general, we need to get better at this throughout cockpit.

In this case it is caused by the code that creates the model.  It resets it to empty and then fills it again incrementally instead of keeping it and applying changes.

I think this is the trigger that makes me try DBusProxies as the model right away.

**Comment by mvollmer** [comment]
> check-servies on rhel-7 needs attention.

Done.

**Comment by stefwalter** [comment]
Fixes #3 woooo

**Comment by stefwalter** [comment]
Fixes #48 ?

**Comment by mvollmer** [comment]
> Fixes #48 ?

Yes.

**Comment by mvollmer** [comment]
> I think this is the trigger that makes me try DBusProxies as the model right away.

Some code here: https://github.com/mvollmer/cockpit/tree/direct-systemd-with-proxies.

It is quite a bit slower than the code here, and in its naive form misses some changes (as expected), and the end result will probably be less code but also less clear than what we have here.  I'll file an issue to follow up on it more seriously, and fix the current code to do smoother updates.

**Comment by mvollmer** [comment]
> The whole overview page is reconstructed incrementally when a service is enabled or disabled.

Done.

**Comment by mvollmer** [comment]
> The "Since" timestamp is the epoch often.

Fixed.

**Comment by mvollmer** [comment]
>  I'll file an issue to follow up on it more seriously

#2162

**Pull request closed without merging**